### DISCUSSION TITLE
Add an shader entry point for ambient light process.

### DISCUSSION THREAD

**Discussion by ODtian** [godotengine/godot-proposals#7530] [3D]
Ambient light contains a lot of information, mainly including sky ambient light and some global illumination information, including light map, voxel gi, sdfgi and ssil.

But currently we cannot read any information about ambient light, if I have a custom brdf, I can apply it to direct lighting (calculated in light()), but I can't apply it to indirect lighting, I can only use the built-in effect in the engine, which causes a visual fragmentation.

Custom handling of this lighting information is useful, and both pbr (replacing the expensive bssrdf with a simple custom brdf for global sss effect) and npr (toon rendering that requires ambient lighting in specific visual style) can make good use of this information.

My suggestion is to add a function entry to the spatial shader to process ambient light information, the position should be after the fragment entry and before the light entry in render pipeline.

**Comment by Calinou** [comment]
- Related to https://github.com/godotengine/godot-proposals/issues/2904.

**Reply by ODtian** [reply]
This issue is more like adding a built-in texture function, but sampling the lightmaps or a light probe.

Add an entry point is another way, you process the exist value and write back.

But all in all, we both want to expose some useful data in godot default render pipeline.

**Comment by ODtian** [comment]
Will, after several trying i realized that what broke my scene is some artifacts caused by bias setting in sdfgi, as i turn it to 0 things looks just as good as voxelgi. it looks physically good enough so there is no need for me to adjust ambient light in a scene shader.

However, for somebody(me included) who wants to refine the ambient light to make their scenes look a little bit better, i will provid a easy implementation within few lines allowing you add some tint or modify ambient light in your ways. And i believe the further implementations will have more data(like gi data) be exposed with 0 cost at runtime.
